### PR TITLE
Canonical AllocID

### PR DESCRIPTION
**Pull request by RalfJung**
This is the Miri side of https://github.com/rust-lang/rust/pull/69408.
This just ports the existing extern statics to the new system; no new shims are added.

Cc @christianpoveda

### CODE REVIEW

File: src/machine.rs [hunk 1]
```diff
@@ -263,32 +287,32 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'tcx> {
...
-    #[inline(always)]
-    fn before_terminator(_ecx: &mut InterpCx<'mir, 'tcx, Self>) -> InterpResult<'tcx> {
-        // We are not interested in detecting loops.
-        Ok(())
+        // Check if we know this one.
+        if let Some(canonical_id) = mem.extra.extern_statics.get(&*link_name) {
+            trace!("canonical_alloc_id: {:?} ({}) -> {:?}", id, link_name, canonical_id);
+            *canonical_id
+        } else {
+            // Return original id; `Memory::get_static_alloc` will throw an error.
```
Comment (oli-obk) [inline src/machine.rs]: An error will only be thrown for AllocIds without a backing allocation

Comment (oli-obk) [reply src/machine.rs]: ah nevermind, this is really just for extern statics

File: src/machine.rs [hunk 2]
```diff
@@ -74,7 +74,11 @@ pub struct MemoryExtra {
     pub stacked_borrows: Option<stacked_borrows::MemoryExtra>,
     pub intptrcast: intptrcast::MemoryExtra,
 
+    /// Mapping extern static names to their canonical allocation.
+    pub(crate) extern_statics: HashMap<&'static str, AllocId>,
```
Comment (eddyb) [inline src/machine.rs]: I wonder if this should be keyed on `Symbol` instead.
Also, does miri not use `FxHashMap`s? They're an easy speed boost for rustc but idk how relevant that is here.

(Sorry if this stuff has been discussed before in the context of miri, I haven't personally seen anything about this)

Comment (RalfJung) [reply src/machine.rs]: Miri doesn't currently use `FxHashMap` but probably should, yeah.

As for `Symbol`, I honestly doubt it's worth all the trouble that brings with it.

Comment (eddyb) [reply src/machine.rs]: > worth all the trouble that brings with it

You always start with `Symbol` for looking up things and then call `.as_str()` on it, the only thing that's more complex is inserting (you have to call `Symbol::intern("...")`) but that happens only once.

**Pull request merged**